src/runtime/virtcontainers/utils/utils.go

```diff
@@ -351,27 +382,61 @@ outer:
 			var waitedPid int
 
 			// "A watched pot never boils" and an unwaited-for process never appears to die!
-			waitedPid, err = syscall.Wait4(pid, &_status, syscall.WNOHANG, &_rusage)
+			waitedPid, err := syscall.Wait4(pid, &_status, syscall.WNOHANG, &_rusage)
 
 			if waitedPid == pid && err == nil {
-				pidRunning = false
-				break outer
+				return false
 			}
 
-			if err = syscall.Kill(pid, syscall.Signal(0)); err != nil {
-				pidRunning = false
-				break outer
+			if isProcessDead(pid) {
+				return false
 			}
 
-			break
-
 		case <-timeout:
-			logger.Warnf("process %v still running after waiting %ds", pid, timeoutSecs)
+			if isProcessDead(pid) {
+				return false
+			} else {
+				logger.Warnf("process %v still running after waiting %ds", pid, timeoutSecs)
+				return true
+			}
+		}
+	}
+}
 
-			break outer
+// WaitLocalProcess waits for the specified process for up to timeoutSecs seconds.
+//
+// Notes:
+//
+// - If the initial signal is zero, the specified process is assumed to be
+//   attempting to stop itself.
+// - If the initial signal is not zero, it will be sent to the process before
+//   checking if it is running.
+// - If the process has not ended after the timeout value, it will be forcibly killed.
+func WaitLocalProcess(pid int, timeoutSecs uint, initialSignal syscall.Signal, logger *logrus.Entry) error {
+	var err error
+
+	// Don't support process groups
+	if pid <= 0 {
+		return errors.New("can only wait for a single process")
+	}
+
+	if initialSignal != syscall.Signal(0) {
+		if err = syscall.Kill(pid, initialSignal); err != nil {
+			if err == syscall.ESRCH {
+				logger.WithField("pid", pid).Warnf("kill encounters ESRCH, process already finished")
+				return nil
+			}
+
+			return fmt.Errorf("Failed to send initial signal %v to process %v: %v", initialSignal, pid, err)
 		}
 	}
 
+	pidRunning, err := waitUsingPidfd(pid, timeoutSecs, logger)
+
+	if err == unix.ENOSYS {
+		pidRunning = waitUsingLegacy(pid, timeoutSecs, logger)
+	}
+
```

Comment:
If the hypervisor isn't daemonized, the shim is the parent of the hypervisor process and it thus **must** `wait4()` or a variant on its child process in some way. That is, the pidfd path must ensure a terminated QEMU child is ripped as well.

When the pidfd points to a child, it is possible to `waitid(P_PIDFD)`. This could be added to `waitUsingPidfd()` but I don't see any API to do that in the go library, apart from `syscall.Syscall6(syscall.SYS_WAITID, ...)`

Reply:
I don't know if this as a problem, as it is right now WaitLocalProcess is used only to wait for that process to exit, not to reap child processes ("init" will reap them after reparenting). There are a two other places in runtime where syscall.Kill with SIGKILL is called and there is no reaping done.

I could try to add reaping for all cases.

Reply:
Fixed, added reaping for pidfd and processes that are killed.`